core/src/main/scala/filodb.core/metadata/Schemas.scala

```diff
@@ -153,8 +158,13 @@ object PartitionSchema {
 
 /**
  * A Schema combines a PartitionSchema with a DataSchema, forming all the columns of a single ingestion record.
+ *
+ * The downsample member is a var because setting it at construction time can potentially result be impossible
+ * when downsample schema is same as raw schema
+ *
+ * Important Note: Serialization will be tricky since there may be loops in object graph. Avoid if possible.
  */
-final case class Schema(partition: PartitionSchema, data: DataSchema, downsample: Option[Schema]) {
+final case class Schema(partition: PartitionSchema, data: DataSchema, var downsample: Option[Schema] = None) {
```

Comment:
Can we please find a way to avoid making downsample a var?  Mutability + case classes is a bad mixture, will lead to initialization bugs and related headaches

Reply:
Please suggest options. This was the best way out. Immutability and cycles in object graphs is not possible unless I am missing something!

Reply:
I don't think there is any problem with having a cycle in the case class, except for how you use it ie the `toString` method.

Reply:
My point is you cant have a cycle with immutable case classes unless you use a var.